masq/src/test_utils/mocks.rs

```diff
@@ -352,9 +393,166 @@ impl TestStreamFactoryHandle {
                     }
                     thread::sleep(Duration::from_millis(100));
                 }
-                Err(_) => break,
+                Err(_) => {
+                    break;
+                }
             }
         }
         accum
     }
 }
+
+#[derive(Clone)]
+pub struct MixingStdout {
+    channel_half: Sender<String>,
+}
+
+impl MixingStdout {
+    pub fn new(sender: Sender<String>) -> Self {
+        MixingStdout {
+            channel_half: sender,
+        }
+    }
+}
+
+impl Write for MixingStdout {
+    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+        self.channel_half
+            .send(std::str::from_utf8(buf).unwrap().to_string())
+            .unwrap();
+        Ok(0)
+    }
+
+    fn flush(&mut self) -> std::io::Result<()> {
+        Ok(())
+    }
+    fn write_fmt(&mut self, fmt: Arguments<'_>) -> std::io::Result<()> {
+        self.channel_half.send(fmt.to_string()).unwrap();
+        Ok(())
+    }
+}
+
+#[derive(Clone)]
+pub struct TerminalPassiveMock {
+    read_line_result: Arc<Mutex<Vec<TerminalEvent>>>,
+}
+
+impl MasqTerminal for TerminalPassiveMock {
+    fn provide_lock(&self) -> Box<dyn WriterLock + '_> {
+        Box::new(WriterInactive {})
+    }
+    fn read_line(&self) -> TerminalEvent {
+        self.read_line_result.lock().unwrap().remove(0)
+    }
+}
+
+impl TerminalPassiveMock {
+    pub fn new() -> Self {
+        Self {
+            read_line_result: Arc::new(Mutex::new(vec![])),
+        }
+    }
+
+    pub fn read_line_result(self, result: TerminalEvent) -> Self {
+        self.read_line_result.lock().unwrap().push(result);
+        self
+    }
+}
+
+//mock incorporating a terminal very similar to the real one, which is run as in-memory though
+
+pub struct TerminalActiveMock {
```

Comment:
Is this still necessary?

Reply:
Yes...I had to say to myself that writing my own mock that'd substitute this one would be too demanding. It's 50/50 with benefits from that.